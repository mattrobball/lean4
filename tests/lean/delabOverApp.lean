/-!
Tests for delaborating over-applied functions.
For example, checking to see that the algorithm for app unexpanders handles over-applications.
-/

/-!
Technically these are not examples of overapplied functions,
but in principle these are two-argument functions that return functions.
-/
def add (f g : Nat → Nat) : Nat → Nat := fun x => f x + g x
def mul (f g : Nat → Nat) : Nat → Nat := fun x => f x * g x

variable (f g : Nat → Nat)

infixl:65 " +' " => add

/-!
Checking the app unexpander that was generated by the `infixl` command works when overapplied.
-/
#check f +' g
#check (f +' g) 1

@[app_unexpander mul]
def unexpandAdd : Lean.PrettyPrinter.Unexpander
  | `($_ $f $g) => `($f * $g)
  | _ => throw ()

/-!
Checking the app unexpander when both applied and overapplied
-/
#check mul f g
#check mul f g 1

/-!
Make sure pp.explicit turns off app unexpanders
-/
section
set_option pp.explicit true
#check mul f g
#check mul f g 1
end
