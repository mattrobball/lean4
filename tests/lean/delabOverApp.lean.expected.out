f +' g : Nat → Nat
(f +' g) 1 : Nat
f * g : Nat → Nat
(f * g) 1 : Nat
mul f g : Nat → Nat
mul f g 1 : Nat
